apps/consulting/posts/improvements-to-bsr-matrix-multiplication-in-pytorch.md

```diff
@@ -0,0 +1,292 @@
+clear
+
+```python
+>>> mat.to_sparse_bsr((2, 2))
+tensor(crow_indices=tensor([0, 3, 4, 6]),
+       col_indices=tensor([0, 1, 2, 1, 0, 2]),
+       values=tensor([[[0, 0],
+                       [1, 0]],
+                        ...
+                      [[1, 0],
+                       [0, 0]]]),
+      size=(6, 6), nnz=6, layout=torch.sparse_bsr)
+```
+
+There are some drawbacks to a block sparse layout, namely we must fully materialize a block if any
+value within it is non-zero. Larger blocks are generally better for vectorized operations, but will
+have more memory overhead.  The example we have used here demonstrates this well. In
+[Figure-1][#fig-1] we have this matrix with non-zero values in yellow, and the 2x2 block size shown
+with the red grid. 
+
+<img src="posts/improvements-to-bsr-matrix-multiplication-in-pytorch/blocksparse_2x2_matplot.png"
+    name="fig-1"
+    alt="A 6x6 grid of colored squares, zero and non-zero values indicated by color. Overlaid is a
+    grid segregating the matrix into 9 2x2 blocks."
+/>
+
+## Matrix Multiplication with Sparse Tensors
+
+In PyTorch, dense matrix multiplication in eager mode will usually forward to a high performance
+math library (i.e. cuBLAS) that implements the [`GEMM`][cublas-docs-gemm]. Once we start to involve
+sparse tensors the situation becomes more complex. We must have a different `GEMM` like operation
+for every pattern of sparse/dense arguments we need to support. Here we will focus on two forms
+involving exactly one sparse argument, we have
+
+- `C(Dense) += A(Sparse) @ B(Dense)` or **DSD**
+- `C(Sparse) += A(Dense) @ B(Dense)` or **SDD**
+
+where `@` is used to represent the matrix multiplication operation.
+
+The DSD pattern is an operation with the same semantics as a dense matrix multiplication. If the
+sparse argument were converted to dense, or all implicit zeros were materialized the result should
+be equivalent. The DSD pattern can be used to compute linear layers with sparse weights. This is an
+function which is semantically equivalent to [`torch.nn.functional.linear`][pytorch-docs-linear]
+
+```python
+def linear(x, W, b=None):
+    tmp = (W @ x.transpose(-2, -1)).transpose(-2, -1)
+    if b is not None:
+        return tmp + b
+    else:
+        return tmp
+```
+
+In this case the weights, `W` can be stored using BSR layout, provided the DSD matrix multiply is
+supported. Note that linear is defined by `x@W^T` and above we have calculated the transpose of this
+expression.
+
+In contrast the SDD pattern does is not semantically equivalent under conversion to dense layout. In
+this case the sparse tensor updated with the result of `A@B` also acts as a mask.  In other words
+`C` will have the same sparsity pattern and that pattern is imposed on the product `A@B`.  In
+PyTorch, we call this operation [`torch.sparse.sampled_addmm`][pytorch-docs-sampled-addmm].  This
+operation is also relevant for transformer models.  It can be used to compute masked gradients for
```

Comment:
We have not said anything else this op is relevant for, so "also" doesn't make much sense here.
```suggestion
operation is particularly relevant for transformer models.  It can be used to compute masked gradients for
```